source/common/stats/thread_local_store.cc

```diff
@@ -201,47 +201,52 @@ void ThreadLocalStoreImpl::releaseScopeCrossThread(ScopeImpl* scope) {
   ASSERT(scopes_.count(scope) == 1);
   scopes_.erase(scope);
 
-  // This is called directly from the ScopeImpl destructor, but we can't delay
-  // the destruction of scope->central_cache_.central_cache_.rejected_stats_
-  // to wait for all the TLS rejected_stats_ caches to be destructed, as those
-  // reference elements of SharedStatNameStorageSet. So simply swap out the set
-  // contents into a local that we can hold onto until the TLS cache is cleared
-  // of all references.
+  // This method is called directly from the ScopeImpl destructor, but we can't
+  // destroy scope->central_cache_ until all the TLS caches are be destroyed, as
+  // the TLS caches reference the Counters and Gauges owned by the central
+  // cache. We don't want the TLS caches to bump the reference-count, as
+  // decrementing the count requires an allocator lock, which would cause a
+  // storm of contention during scope destruction.
   //
-  // We use a raw pointer here as it's easier to capture it in the lambda.
-  auto rejected_stats = new StatNameStorageSet;
-  rejected_stats->swap(scope->central_cache_.rejected_stats_);
+  // So instead we have a 2-phase destroy:
+  //   1. destroy all the TLS caches
+  //   2. destroy the central cache.
+  //
+  // Since this is called from ScopeImpl's destructor, we must move the central
+  // cache storage out of the ScopeImpl and into an explicitly managed
+  // pointer. We use a raw pointer here as it's easier to capture it in the
+  // lambda.
```

Comment:
Although not explicitly needed, can we just use a shared_ptr here? I think it would simplify a lot of this logic as it can be captured in the lambda, you wouldn't need an additional allocation and can just move the shared_ptr out of the scope, etc.

Reply:
Good idea about using a shared_ptr here. Note that storing it as a shared_ptr in the ScopeImpl will increase memory usage slightly, requiring regolding some expected-memory values from my memory-tests, but I think it will not be material and I should probably not worry about it :)

Reply:
...and the reason -- I'd guess -- that you originally used a shared_ptr was so that you didn't have to wait for the TLS cache objects to be deleted before deleting the central cache.

I added the after-all-TLS-caches-are-removed callback myself sometime back to deal with the storage for the cache of StatNames for rejected stats. At that time I could've just done this.

Reply:
...done. That actually simplified a lot of code, though I'm a bit nervous about the timing of when the CentralCacheEntry destructor gets run in obscure scenarios. If it gets run after the destruction of the SymbolTable, *and* there are rejected stats, it may crash. But this probably only happens when an exception prevents startup.

The advantage of maintaining the pointer directly, as I had before this most recent commit, is that we force cleanup of the rejected-stat-name table on destruction of the ThreadLocalStore, and generally the symbol-table will outlast that.

I think what we have is OK as long as we ensure the symbol-tables outlast the dispatchers.